spec/app/ics-029-fee-payment/README.md

```diff
@@ -0,0 +1,391 @@
+    // pay the reverse fee to the reverse relayer address
+    // refund extra tokens to original fee payer(s)
+}
+
+// PayFee is a callback implemented by fee module called by the ICS-4 TimeoutPacket handler.
+function PayTimeoutFee(packet: Packet, timeout_relayer: string) {
+    // pay the timeout fee to the timeout relayer address
+    // refund extra tokens to original fee payer(s)
+}
+```
+
+
+The fee module should also expose the following queries so that relayers may query their expected fee:
+
+```typescript
+// Gets the fee expected for submitting ReceivePacket msg for the given packet
+// Caller should provide the intended relayer address in case the fee is dependent on specific relayer(s).
+function GetReceiveFee(portID, channelID, sequence, relayer) Fee
+
+// Gets the fee expected for submitting AcknowledgePacket msg for the given packet
+// Caller should provide the intended relayer address in case the fee is dependent on specific relayer(s).
+function GetAckFee(portID, channelID, sequence, relayer) Fee
+
+// Gets the fee expected for submitting TimeoutPacket msg for the given packet
+// Caller should provide the intended relayer address in case the fee is dependent on specific relayer(s).
+function GetTimeoutFee(portID, channelID, sequence, relayer) Fee
+```
+
+Since different chains may have different representations for fungible tokens and this information is not being sent to other chains; this ICS does not specify a particular representation for the `Fee`. Each chain may choose its own representation, it is incumbent on relayers to interpret the Fee correctly.
+
+### IBC Module Wrapper
+
+The fee module will implement its own ICS-26 callbacks that wrap the application-specific module callbacks. This fee module middleware will ensure that the counterparty module supports incentivization and will implement all fee-specific logic. It will then pass on the request to the embedded application module for further callback processing.
+
+In this way, custom fee-handling logic can be hooked up to the IBC packet flow logic without placing the code in the ICS-4 handlers or the application code. This is valuable since the ICS-4 handlers should only be concerned with IBC correctness, and the application handlers should not be handling fee logic that is universal amongst all other incentivized applications. In fact, a given application module should be able to be hooked up to any fee module with no further changes to the application itself.
+
+As mentioned above, the fee module will implement the ICS-26 callbacks, and can embed an application IBC Module. All ICS-26 callbacks in the fee module will call into the embedded application's callback. Thus, only the fee callbacks that do something additional are explicitly specified here.
+
+#### Fee Protocol Negotiation
+
+The fee middleware will negotiate its fee protocol version with the counterparty module by prepending its own version to the application version. 
```

Comment:
This seems a little awkward to me, although workable - should we encode a list of strings instead in some standard way?

Reply:
Do you mean changing channel version to be a channel list?

I'm in favor, though this would be a breaking change, so something to do in future.

If you mean creating a list of strings inside the string that gets deserialized in a particular way, that is what my proposal is doing. Though there may be a much better way to do it.

Reply:
I think this is fine for now.